### PR TITLE
Add CertificateRequest username to CEL Validator with serviceaccount functions

### PR DESCRIPTION
**Pull request by jamesglennan**
Closes https://github.com/cert-manager/approver-policy/issues/506

This PR is to add the username associated with a certificate request into the CEL Validator, so it is possible to approve or deny certificate requests based on who/what requested them. This is useful specifically for validating SPIFFE SVIDs where the URI field is in the format of `spiffe://trust.domain/ns/serviceAccountNamespace/sa/serviceAccountName`

### CODE REVIEW

**Comment by SgtCoDFish** [comment]
/ok-to-test

(I don't think tests will run on a draft PR, but at least now they'll run when this is marked ready for review!)

**Comment by SgtCoDFish** [comment]
I think this looks like a great start. I had a brief review - this isn't really my area of expertise but it looks how I'd expect this to look! 👍

File: pkg/internal/approver/validation/validator_test.go
Comment (erikgb) [inline pkg/internal/approver/validation/validator_test.go:100]:
```suggestion
	v := &validator{expression: "isServiceAccount(cr.username) && self.startsWith('spiffe://acme.com/ns/%s/sa/%s'.format([serviceAccount(cr.username).getNamespace(),serviceAccount(cr.username).getName()]))"}
```

I would suggest a slightly modified API to this new feature. Please let me know what you think and if it is possible to implement. I will await a review of the code/implementation until we are aligned on how this will will appear for the user. 😺

**Review by erikgb — CHANGES_REQUESTED** [review]
This looks great! Thanks @jamesglennan! I've done an initial review, and I like the way this is heading!

/approve
/hold

File: pkg/internal/approver/validation/validator_test.go
Comment (jamesglennan) [reply pkg/internal/approver/validation/validator_test.go]: Added you changes and modified the tests- I also havent had a chance to test this in a real cluster yet

File: pkg/internal/approver/validation/serviceaccount.go
Comment (erikgb) [inline pkg/internal/approver/validation/serviceaccount.go:49]: Is this correct? I would assume this should return a string. Maybe add a `String()` function to ServiceAccount type, returning the standard Kubernetes SA string: `system:serviceaccount:<namespace>:<name>`? WDYT?

File: pkg/internal/approver/validation/serviceaccount.go
Comment (erikgb) [inline pkg/internal/approver/validation/serviceaccount.go:63]: Looking at the upstream URL type, this function looks a bit different. Why?

https://github.com/kubernetes/apiserver/blob/1ea7befac323763d28173ed613a9e73ef1ef9c16/pkg/cel/url.go#L53-L61

File: pkg/internal/approver/validation/serviceaccount.go
Comment (erikgb) [inline pkg/internal/approver/validation/serviceaccount.go:67]:
```suggestion
	otherSA, ok := other.(ServiceAccount)
```

Or does `otherDur` have a special meaning I don't understand? 😆

File: pkg/internal/approver/validation/serviceaccount.go
Comment (erikgb) [inline pkg/internal/approver/validation/serviceaccount.go:108]: Shouldn't we return an error here instead? An error here will be extremely hard to debug with the current lenient handling.

**Review by erikgb — COMMENTED** [review]
Sorry for the delayed review here! 😊 I have added some comments and suggestions. Hoping we can get this merged soon!

File: pkg/internal/approver/validation/serviceaccount.go
Comment (jamesglennan) [reply pkg/internal/approver/validation/serviceaccount.go:63]: Its the same, the difference is that in the URL library `types.NewTypeValue("kubernetes.URL")` is deprecated, so i used the new function.

If I'm honest I'm not sure what the distinction between

```
	URLObject = decls.NewObjectType("kubernetes.URL")
	typeValue = types.NewTypeValue("kubernetes.URL")
	URLType   = cel.ObjectType("kubernetes.URL")
```

They all seem to return the same thing?

File: pkg/internal/approver/validation/serviceaccount.go
Comment (jamesglennan) [reply pkg/internal/approver/validation/serviceaccount.go:33]: I'm not sure what you mean by make the type as receiver? The saLib struct needs methods to compile the new CEL functions and the ServiceAccount struct is to store and convert back and forth

Comment (erikgb) [reply pkg/internal/approver/validation/serviceaccount.go:33]: AFAIK, it's common to keep the methods on a struct type after the struct definition in Golang. I am addressing these methods: https://github.com/cert-manager/approver-policy/blob/5ccd683693ff056880eb1388a33f2b6f13d0c688/pkg/internal/approver/validation/serviceaccount.go#L150-L160

https://go.dev/tour/methods/1

File: pkg/internal/approver/validation/serviceaccount.go
Comment (erikgb) [reply pkg/internal/approver/validation/serviceaccount.go:63]: Ok, thanks!

**Review by SgtCoDFish — APPROVED** [review]
/unhold
/lgtm
/approve

Thanks so much for this James! I agree with Erik that this looks great, and we discussed his nit in standup today and it's not a blocker.

Intent here is to release an alpha version of approver-policy for you to test and then if that works for you after a while we'll release it properly.

First step is to merge, so I'll do that now :D

**Review by erikgb — APPROVED** [review]

**Pull request merged**
